### PR TITLE
DFT torch model

### PR DESCRIPTION
**Pull request by advikavs**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by rbharath — COMMENTED** [review]
At a high level this all looks reasonable, but it's hard to do a review until docs/doctests/tests are added in. Can you go ahead and flesh out the PR with those? I will do a thorough review then

File: deepchem/data/data_loader.py [hunk 1]
```diff
@@ -1690,15 +1690,17 @@ def create_dataset(self,
 
         def shard_generator():
             entries = self._get_shards(inputs)
+            i = 0
             for shard in entries:
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1694]: You can instead do 
```
for i, shard in enumerate(entries)
  ...
```
to avoid having to manually update `i`

File: deepchem/data/data_loader.py [hunk 2]
```diff
@@ -1740,4 +1742,4 @@ def _featurize_shard(self, shard):
                 "Unknown key in yaml file. Please check format for correctness."
             )
         x = DFTEntry.create(e_type, true_val, systems)
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1744]: Why is only 1 entry being created for an entire shard? This step should return as many DFTEntry objects as there were datapoints in the original shard

File: deepchem/models/dft/dftxc.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import torch
+from deepchem.feat.dft_data import DFTEntry, DFTSystem
+from deepchem.models.dft.nnxc import HybridXC
+from deepchem.models.losses import Loss, L2Loss
+from deepchem.models.torch_models.torch_model import TorchModel
+from typing import List, Optional, Tuple, Any
+import numpy as np
+
+
+class DFTXC(torch.nn.Module):
```
Comment (rbharath) [inline deepchem/models/dft/dftxc.py:12]: Reminder for docstring/usage example

File: deepchem/models/dft/dftxc.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        hybridxc = HybridXC("lda_x", self.model, aweight0=0.0)
+        for entry in inputs:
+            evl = XCNNSCF(hybridxc, entry)
+            qcs = []
+            for system in entry.get_systems():
+                qcs.append(evl.run(system))
+            return [(entry.get_val(qcs))]
+
+
+class XCModel(TorchModel):
```
Comment (rbharath) [inline deepchem/models/dft/dftxc.py:35]: Reminder for docstring/usage example

File: deepchem/models/dft/dftxc.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+        inputs, labels, weights = batch
+        labels = [torch.from_numpy(inputs[0][0].get_true_val())]
+        labels[0].requires_grad_()
+        w = np.array([1.0])
+        weights = [torch.from_numpy(w)]
+        return (inputs, labels, weights)
+
+
+class ExpM1Activation(torch.nn.Module):
```
Comment (rbharath) [inline deepchem/models/dft/dftxc.py:69]: Reminder for docstring/usage example

File: deepchem/models/dft/dftxc.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return torch.exp(x) - 1
+
+
+def construct_nn_model(ninp: int, nhid: int, ndepths: int, modeltype: int):
```
Comment (rbharath) [inline deepchem/models/dft/dftxc.py:75]: We need a docstring. We may want to make this a private function

File: deepchem/models/tests/test_dftxc.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+
+@pytest.mark.dqc
+def test_dftxcloss():
+    inputs = 'deepchem/models/tests/assets/test_dftxcdata.yaml'
+    k = DFTYamlLoader()
+    dataset = (k.create_dataset(inputs))
+    model = XCModel("lda_x", batch_size=1)
+    loss = model.fit(dataset, nb_epoch=1, checkpoint_interval=1)
+    assert loss == 0.003224595397314571
```
Comment (rbharath) [inline deepchem/models/tests/test_dftxc.py:13]: I'd recommend instead checking that the loss goes down as you train more epochs. Checking explicit loss values is usually not numerically stable

**Review by rbharath — COMMENTED** [review]
Did a first preliminary review. Take a look at my comment on `_featurize_shard`. I think the shape of the output may be possibly wrong

File: deepchem/models/tests/test_dftxc.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+                           model_dir=model_dir)
+    reload_model.restore()
+    predict = model.predict(dataset)
+    r_predict = reload_model.predict(dataset)
+    assert np.all(predict == r_predict)
+    metric = dc.metrics.Metric(dc.metrics.mae_score)
+    scores = model.evaluate(dataset, [metric])
+    assert scores['mae_score'] < 0.3
+
+
```
Comment (advikavs) [inline deepchem/models/tests/test_dftxc.py:37]: add pytest mark

**Comment by advikavs** [comment]
Reopened in #3363
